File: ios/Packages/GemstoneServices/Sources/Stores/PriceAlertStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemPriceAlertStore
import struct Gemstone.PriceAlert
import class Gemstone.PriceAlertFormatter
import GemstonePrimitives
import Primitives
import Store

public final class GemstonePriceAlertStore: GemPriceAlertStore, @unchecked Sendable {
    private let store: PriceAlertStore

    public init(store: PriceAlertStore) {
        self.store = store
    }

    public func getPriceAlerts(assetId: String?) async throws -> [Gemstone.PriceAlert] {
        let alerts = try assetId.map { try store.getPriceAlerts(for: $0) } ?? store.getPriceAlerts()
        return alerts.map { $0.map() }
    }

    public func updatePriceAlerts(alerts: [Gemstone.PriceAlert], deleteIds: [String]) async throws {
        try store.diffPriceAlerts(
            deleteIds: deleteIds,
            alerts: alerts.map { (id: PriceAlertFormatter.shared.alertId(alert: $0), alert: $0.map()) },
        )
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/PriceAlertStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.PriceAlertsDao
import com.gemwallet.android.data.services.store.database.entities.toPriceAlert
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import kotlinx.coroutines.flow.map
import uniffi.gemstone.GemPriceAlertStore
import uniffi.gemstone.PriceAlertFormatter

class GemstonePriceAlertStore(private val priceAlertsDao: PriceAlertsDao, private val priceAlertFormatter: PriceAlertFormatter) : GemPriceAlertStore {

    override suspend fun getPriceAlerts(assetId: String?): List<uniffi.gemstone.PriceAlert> {
        val records = assetId?.let { priceAlertsDao.getAllPriceAlerts(it) } ?: priceAlertsDao.getAllPriceAlerts()
        return records.map { it.toPriceAlert().toGem() }
    }

    override suspend fun updatePriceAlerts(alerts: List<uniffi.gemstone.PriceAlert>, deleteIds: List<String>) {
        priceAlertsDao.update(alerts.map { it.toPrimitives().toRecord(priceAlertFormatter.alertId(it)) }, deleteIds)
    }
}
